Pcr primer design for dna assembly

DmitriiSukhomlinov opened the pull request:
XML tests have passed (Windows, Linux, macOS), GUI tests have passed (Linux).
[Documentation](https://doc.ugene.net/wiki/display/UM/PCR+Primer+Design+for+DNA+Assembly)
[Plans](https://docs.google.com/document/d/1upVVUEUAYjlVlHGb2WotCpqMHCsJ8JpdVezAqYT8f5M/edit?usp=sharing)

DmitriiSukhomlinov committed:
UGENE-7244 Create PCR Primer Design for DNA assembly plugin (#401)

* UGENE-7244 Create PCR Primer Design for DNA assembly plugin
UGENE-7245 Create a tab is the Sequence View

DmitriiSukhomlinov committed:
UGENE-7275 (#413)

* UGENE-7275 preparing part of the algorithm

* UGENE-7275 update exclude values

evelinabis committed:
UGENE-7321. Increase width of option panel to make the tab fully (#417)

visible.
Min option tab width increased.

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

rasputinkirill committed:
UGENE-7278 Process algorithm results (#425)

DmitriiSukhomlinov committed:
UGENE-7275 Implement algorithm (#427)

* UGENE-7275 continue making the algorithm

* UGENE-7275 hetero dimers

* UGENE-7275 fix windows size and remove result locker

DmitriiSukhomlinov committed:
UGENE-7275 forgotten files

rasputinkirill committed:
UGENE-7331 Don't create a new annotation table, just add a new group to the exist one (#437)

rasputinkirill committed:
UGENE-7330 Crash on the second search (#435)

rasputinkirill committed:
UGENE-7329 Crash when close option the panel tab (#434)

evelinabis committed:
UGENE-7321. Increase width of option panel to make the tab fully visible 1 (#444)

* UGENE-7321. Increase width of option panel to make the tab fully
visible.
Min option tab width increased.

* UGENE-7321. Increase width of option panel to make the tab fully
visible.

Min option tab width increased. 1

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

DmitriiSukhomlinov committed:
UGENE-7275 (#447)

* UGENE-7275 continue the algorithm implementing

* UGENE-7275

* UGENE-7275 check forward only

* UGENE-7275 add selection on select manually

rasputinkirill committed:
UGENE-7332 Make result table visible if nothing has been found (#449)

evelinabis committed:
UGENE-7325. Make the "PCR primer design" tab visible for DNA sequence only (#443)

* UGENE-7325. Make the "PCR primer design" tab visible for DNA sequence
only.
The widget is dynamically enabled/disabled when selecting a
nucleotide/non-nucleotide sequence.
Reconstructed UI file: added a widget to the root containing other
widgets, added a warning label.

* UGENE-7325. Make the "PCR primer design" tab visible for DNA sequence
only.
Registered PCR Primer Design tab in GUITest enum.
Added GUI test.

* UGENE-7325. Make the "PCR primer design" tab visible for DNA sequence
only.

Added the ability to filter option panel tabs by RNA.
PCR Primer Design tab is available only for sequence files containing
DNA sequence.
Added GUI test for new RNA behavior.
Simplified test_7325.
Minor changes: added newline at end of file.

* UGENE-7325. Make the "PCR primer design" tab visible for DNA sequence
only.

Changes by review.
Small code refactoring, such as moving and splitting code into separate
methods.
Fixed region manually selection.
Fixed scenario from pull request.
Added a change allowed spinboxes ranges when changing sequence length.
Added Select manually button disabling when changing active sequence.

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

evelinabis committed:
UGENE-7311. Implement "FindPresenceOfUnwantedParametersTask" (#454)

* UGENE-7311. Implement "FindPresenceOfUnwantedParametersTask".
Added search for homo- and heterodimers in
FindPresenceOfUnwantedParametersTask.
Created a dialog displaying unwanted structures.
Added logic for processing FindPresenceOfUnwantedParametersTask results
and user responses.

* UGENE-7311. Implement "FindPresenceOfUnwantedParametersTask".
Minor changes for beautiful output.

* UGENE-7311. Implement "FindPresenceOfUnwantedParametersTask".
Changes by review.

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

DmitriiSukhomlinov committed:
UGENE-7275 minor improvements

DmitriiSukhomlinov committed:
macOs build fix

DmitriiSukhomlinov committed:
add comment

evelinabis committed:
UGENE-7347. Add dimer melting temperature to self- and hetero- dimer reports. (#468)

* UGENE-7347. Add dimer melting temperature to self- and hetero- dimer
reports.

* UGENE-7347. Add dimer melting temperature to self- and hetero- dimer
reports.
Added forgotten files.

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

rasputinkirill committed:
UGENE-7336 Export the result sequence with the backbone sequence on double click (#466)

evelinabis committed:
UGENE-7348. Create notification report (#479)

* UGENE-7348. Create notification report.

* UGENE-7348. Create notification report.

Remove html font styles.

* UGENE-7348. Create notification report.
Changes by review.

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

rasputinkirill committed:
UGENE-7336 Export the result sequence with the backbone sequence on double click (#497)

rasputinkirill committed:
UGENE-7345 Crash if click on the element in the result table, but the Annotation table'h been removed (#496)

DmitriiSukhomlinov committed:
update release-bundle.txt

DmitriiSukhomlinov committed:
update release-bundle.txt linux and mac

DmitriiSukhomlinov committed:
translations update

DmitriiSukhomlinov committed:
UGENE-7374 Forward and reverse primers shouldn't have unwanted hetero-dimers (#510)

* UGENE-7374 algorithm rewriting

* UGENE-7374 update comments

DmitriiSukhomlinov committed:
UGENE-7373 Add generate sequence feature (#533)

rasputinkirill committed:
UGENE-7335 SAFE_POINT after confirmation in "File Modification Detected" dialog (#534)

evelinabis committed:
Remove duplicate directive include (#543)

evelinabis committed:
UGENE-7375. Check user forward and reverse primers. (#538)

* UGENE-7375. Check user forward and reverse primers.
Added user primers to report.
Changes by review.
Splited "Unwanted connections of user primers" table into two.

* UGENE-7375. Check user forward and reverse primers.
Refactored.

* UGENE-7375. Check user forward and reverse primers.
Namespace ReportUtils moved to struct PCRPrimerDesignTaskReportUtils in
separate files.

* UGENE-7375. Check user forward and reverse primers.
Changes by review and other changes.

* Update PCRPrimerDesignTaskReportUtils.h

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

DmitriiSukhomlinov committed:
remove debug lines

DmitriiSukhomlinov committed:
General updates

DmitriiSukhomlinov committed:
Optimize calculation loop

DmitriiSukhomlinov committed:
Remove primer library from U2Core and move it back to pcr

evelinabis committed:
Update PCRPrimerDesignForDNAAssemblyOPWidget.ui (#582)

DmitriiSukhomlinov committed:
icon

evelinabis committed:
UGENE-7442. The reverse sequences in Task report [PCR Primer Design For DNA Assembly Task] should be taken from complementary sequence (#601)

* UGENE-7432. Implement GUI tests.
First partial implementation.

* UGENE-7432. Implement GUI tests.

1. Increased the width of the PCR Primer Design tab.
2. Increased the width of the Color column in the Annotations
   Highlighting tab.
3. Corrected tests related to changes 1-2.
4. Refactored PCR Primer Design utils.
5. Refactored PCR Primer Design tests.
6. Added missing PCR Primer Design tests.

* UGENE-7432. Implement GUI tests.
Build fixed.

* UGENE-7432. Implement GUI tests

Minor changes.

* UGENE-7432. Implement GUI tests.

Changes by review.
Return to the previous state of the options panel. Return covers the
tests too.
Adding line breaks at the ends of files.

* UGENE-7432. Implement GUI tests.

Minor readability changes.

* UGENE-7432. Implement GUI tests.

Removed changes related to increasing Option Panel width.

* UGENE-7432. Implement GUI tests.

Reformat with the clangformat style.

* UGENE-7432. Implement GUI tests.

Removed consts for locals (by review).

* UGENE-7442. The reverse sequences in Task report [PCR Primer Design For
DNA Assembly Task] should be taken from complementary sequence.

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

evelinabis committed:
UGENE-7321. Increase width of option panel to make the tab fully visible (#602)

* UGENE-7432. Implement GUI tests.
First partial implementation.

* UGENE-7432. Implement GUI tests.

1. Increased the width of the PCR Primer Design tab.
2. Increased the width of the Color column in the Annotations
   Highlighting tab.
3. Corrected tests related to changes 1-2.
4. Refactored PCR Primer Design utils.
5. Refactored PCR Primer Design tests.
6. Added missing PCR Primer Design tests.

* UGENE-7432. Implement GUI tests.
Build fixed.

* UGENE-7432. Implement GUI tests

Minor changes.

* UGENE-7432. Implement GUI tests.

Changes by review.
Return to the previous state of the options panel. Return covers the
tests too.
Adding line breaks at the ends of files.

* UGENE-7432. Implement GUI tests.

Minor readability changes.

* UGENE-7432. Implement GUI tests.

Removed changes related to increasing Option Panel width.

* UGENE-7432. Implement GUI tests.

Reformat with the clangformat style.

* UGENE-7432. Implement GUI tests.

Removed consts for locals (by review).

* UGENE-7321. Increase width of option panel to make the tab fully
visible.

Minimum width of the "PCR Primer Design for DNA assembly" tab increased.
Changed behavior of OptionsScrollArea::sizeHint.
Added GUI test.

* UGENE-7432. Implement GUI tests.

Comments indentation is unified.

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

evelinabis committed:
UGENE-7432. Implement GUI tests (#605)

* UGENE-7432. Implement GUI tests.

1. Increased the width of the PCR Primer Design tab.
2. Increased the width of the Color column in the Annotations
   Highlighting tab.
3. Corrected tests related to changes 1-2.
4. Refactored PCR Primer Design utils.
5. Refactored PCR Primer Design tests.
6. Added missing PCR Primer Design tests.

* UGENE-7432. Implement GUI tests.

Changes by review.
Return to the previous state of the options panel. Return covers the
tests too.
Adding line breaks at the ends of files.

* UGENE-7432. Implement GUI tests.

Removed changes related to increasing Option Panel width.

* UGENE-7432. Implement GUI tests.

Formatted with clang.
Changes by review.
Decreased tab min width.

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

Commit:
Implement PCR Primer Design GUI tests. Fix tab_0002 test.

DmitriiSukhomlinov committed:
fix pcr primer design tab

DmitriiSukhomlinov committed:
UGENE-7467 Create the Russian translations (#609)

DmitriiSukhomlinov committed:
restore deleted files

rasputinkirill committed:
UGENE-7431 Implement XML tests (#604)

Commit:
Fixed GUI tests after reorganizing the
PCRPrimerDesignForDNAAssemblyOPWidget.

evelinabis committed:
UGENE-7444. After reopening the PCR Primer Design tab, the "Select manually" button is pressed, but should not be (#606)

* UGENE-7444. After reopening the PCR Primer Design tab, the "Select
manually" button is pressed, but should not be.

Fixed.
Added GUI test.

* UGENE-7444. After reopening the PCR Primer Design tab, the "Select
manually" button is pressed, but should not be.

Changes by review: the crutch is removed. Changes from
PCRPrimerDesignForDNAAssemblyOPWidget moved to
PCRPrimerDesignForDNAAssemblyOPSavableTab.

* UGENE-7444. After reopening the PCR Primer Design tab, the "Select
manually" button is pressed, but should not be.

Changes by review: made as it's done several lines above.

Co-authored-by: Evelina Biserova <biserova@unipro.ru>

yalgaer commented on src/corelibs/U2Gui/src/options_panel/OPWidgetFactory.h:
```diff
@@ -69,6 +71,8 @@ class U2GUI_EXPORT OPFactoryFilterVisitorInterface {
     virtual bool alphabetPass(DNAAlphabetType factoryAlphabetType) = 0;
     virtual bool atLeastOneAlphabetPass(DNAAlphabetType factoryAlphabetType) = 0;
+    // Returns true, if there is a DNA alphabet (default or extended).
+    virtual bool atLeastOneDnaPass() const = 0;
```
I already commented about this change 2-3 months ago: why do we need this method if we have already have  'atLeastOneAlphabetPass'?

evelinabis replied:
Please see behavior for _common_data/fasta/RNA.fa: there is no PCR Primer Design tab, but there is In Silico PCR tab which is useless for current sequences

yalgaer replied:
Could you please provide a details how "atLeastOneAlphabetPass" is different from "atLeastOneDnaPass" in case if you want to match DNA sequences only? I think "atLeastOneAlphabetPass" will do the job

evelinabis replied:
`atLeastOneAlphabetPass` approves files that contain one or more sequences with an alphabet defined as `DNAAlphabetType`.
`atLeastOneDnaPass` approves files that contain one or more sequences with an alphabet identified as DNA or extended DNA.

yalgaer replied:
I see now: DNAAlphabetType does not differentiate between RNA and DNA. You are right. In this case I would use a serializable alphabet-id as the filer, not a runtime pointer. You can rework it as suggested or resolve the issue in the current state.

evelinabis replied:
@yalgaer, please see the [changes](https://github.com/ugeneunipro/ugene/pull/619/commits/55c0760e9e073fe881410001c15049e817d1a613): is this what you wrote about?

yalgaer replied:
Yes, thank you.

yalgaer commented on etc/script/linux/release-bundle.txt:
```diff
@@ -1350,6 +1350,7 @@ plugins/libngs_reads_classification.so
 plugins/libopencl_support.so
 plugins/liborf_marker.so
 plugins/libpcr.so
+plugins/libpcr_primer_design_for_dna_assembly.so
```
- Merge to "master" requires a good written code and no crashes/unexpected UI behavior.
- Bundling into the release requires a comparison with other tools and some proof that our algorithm works at least as good as others. I suggest removing "release-bundle.txt" changes - without it  you can merge this PR much faster.

DmitriiSukhomlinov replied:
Don't get the idea at all. If we remove these changes, we won't be able even to build a package with this plugin

yalgaer replied:
We can merge a big chunk of code into the master and this will be a big win. All next patches will be small a manageable.

To build the "release" bundle you will still need a separate branch with a very trivial changes in release-bundle.txt.
Note that release-bundle.txt is only used in the release, and not used during the build or testing process.

DmitriiSukhomlinov replied:
Ok, I'd ask more direct - what do you want me to do exactly?

yalgaer replied:
Remove changes in release-bundle.txt from the PR - merging this PR into the master and inclusion of this plugin into a public release are 2 different stages.

yalgaer commented on tests/pcr_dna_assembly/test_0001.xml:
```diff
@@ -0,0 +1,20 @@
+    insertTo="5"
+    leftArea="0..70"
+    rightArea="354..71"
+    expectedResultsList="0..0;0..0;0..0;0..0;0..0;0..0;0..0;0..0"
```
What does this result mean? No results found?
Please add a comment into each test what does it test and what does the result means.

DmitriiSukhomlinov replied:
@rasputinkirill please, fix that

rasputinkirill replied:
Comment added

yalgaer commented on tests/pcr_dna_assembly/test_0005.xml:
```diff
@@ -0,0 +1,21 @@
+    otherSequencesInPcrUrl="pcr_primer_design/other_in_pcr.fa"
+    expectedResultsList="28..22;336..22;0..0;0..0;0..0;0..0;0..0;0..0"
+    />
+
```
I see only a couple of tests for meltingPoint and backboneSequence. 
What about self and hetero-dimers or  other important params?

DmitriiSukhomlinov replied:
Don't have a good data for them. I can kinda make them, but this requires time, I do not have it now

yalgaer replied:
Ok it is not a blocker for merge into the master

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/utils/UnwantedConnectionsUtils.h:
```diff
@@ -0,0 +1,72 @@
+public:
+    UnwantedConnectionsUtils() = delete;
+
+    static bool isUnwantedSelfDimer(const QByteArray& forwardSequence,
```
Please document all public API.

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/corelibs/U2Gui/src/options_panel/OptionsPanelWidget.cpp:
```diff
@@ -164,6 +169,27 @@ GroupOptionsWidget *OptionsPanelWidget::createOptionsWidget(const QString &group
+            QMargins margins = curWidget->contentsMargins();
+            marginSum += margins.left() + margins.right();
+        }
+        if (optionsScrollArea != nullptr) {
```
When optionsScrollArea can be nullptr?

DmitriiSukhomlinov replied:
@EvelinaBiserova as far as you did this, could you answer, please?

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/corelibs/U2Gui/src/options_panel/OptionsPanelWidget.h:
```diff
@@ -42,6 +42,13 @@ class OptionsScrollArea : public QScrollArea {
 
     /** Ensures that the scroll area would have an appropriate width */
     virtual QSize sizeHint() const;
```
add override

DmitriiSukhomlinov replied:
OK

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPSavableTab.cpp:
```diff
@@ -0,0 +1,81 @@
+}
+
+QVariant PCRPrimerDesignForDNAAssemblyOPSavableTab::getChildValue(const QString& childId) const {
+    ResultTable* productTable = qobject_cast<ResultTable*>(getChildWidgetById(childId));
```
nit: do not write Type twice, use auto. Same for other places in the patch.

DmitriiSukhomlinov replied:
OK

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+set(UGENE_PLUGIN_NAME pcr_primer_design_for_dna_assembly)
```
Register in the top-level CMake file

DmitriiSukhomlinov replied:
Don't get the idea

yalgaer replied:
Add 
`add_subdirectory(src/plugins/pcr_primer_design_for_dna_assembly)`

DmitriiSukhomlinov replied:
Done

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+
+    auto seqLength = annDnaView->getActiveSequenceContext()->getSequenceLength();
+
+    // Default values, have been provided by the customer
```
This looks very unsafe for large sequences. May be limit the regions by length?
Large sequences >2GB (like a whole human genome) won't work correctly with QT spinboxes.

DmitriiSukhomlinov replied:
What do you want me to do about it?

yalgaer replied:
Check that the options panel is enabled only for sequences < 2Gb. In this case you can safely cast int64 to int32 inside of your code.

DmitriiSukhomlinov replied:
Are we have some examples of this?

yalgaer replied:
You can always generate one in Random Sequence Generator or download some file with contigs from sequencer and open them merged. Usually such files are 10Gb+ in size

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+    connect(pbAddToReverse5, &QAbstractButton::clicked, this, &PCRPrimerDesignForDNAAssemblyOPWidget::sl_add5ReverseSequence);
+    connect(pbAddToReverse3, &QAbstractButton::clicked, this, &PCRPrimerDesignForDNAAssemblyOPWidget::sl_add3ReverseSequence);
+
+    leFilter->setValidator(new PrimerValidator(this, false));
```
PrimerValidator does not accept extended alphabet, but still allows 'N' symbol. Do you want to allow 'unknowns' in primer sequences?

DmitriiSukhomlinov replied:
Let allow the to be, but I do not know now, actually. This has to be questioned to the customer. He is very busy at the moment and so I do not want to wait for the response. [The issue](https://ugene.dev/tracker/browse/UGENE-7477)

yalgaer replied:
Ok, this does not block merge into the master but should be resolved before the public release.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+    settings.bachbone5Length = sbBackbone5Length->value();
+    settings.bachbone3Length = sbBackbone3Length->value();
+
+    settings.leftArea.startPos = (int)(sbLeftAreaStart->value()) - 1;
```
nit: cast from int to int

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+    SAFE_POINT(activeSequenceContext != nullptr, L10N::nullPointerError("ADVSequenceObjectContext"), );
+
+    U2SequenceObject* sequenceObject = activeSequenceContext->getSequenceObject();
+    SAFE_POINT(NULL != sequenceObject, L10N::nullPointerError("Sequence Object"), );
```
Impossible situation: sequence context always has sequence object

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+    SAFE_POINT(NULL != sequenceObject, L10N::nullPointerError("Sequence Object"), );
+
+    U2OpStatus2Log os;
+    auto sequence = sequenceObject->getWholeSequenceData(os);
```
unsafe. Check that primers plugin code does not run for very large sequence. This way you will fix a lot of int64 issues  in the plugin's code.

DmitriiSukhomlinov replied:
Is there an example of such thing?

yalgaer replied:
Any genome sequencer results or whole genome sequences like http://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/

DmitriiSukhomlinov replied:
I am able to find a large sequence) I mean, the example of the limitation code

yalgaer replied:
I see the "limitation" code here: `sequenceObject->getWholeSequenceData(os);`

If the sequence object is big enough and you are searching for primers in some limited 1k area the whole sequence will be copied. For 2GB+ data it simply won't work. For smaller sequences it will work inefficiently. We have a special class for that SequenceWalker that processes sequence of any size by chunks.

You can avoid the complex code of SequenceWalker by enabling the new plugin functionality only small sequences (up to several megabytes)

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+
+    pcrTask = new PCRPrimerDesignForDNAAssemblyTask(settings, sequence);
+    auto ts = AppContext::getTaskScheduler();
+    SAFE_POINT(ts != nullptr, L10N::nullPointerError("TaskScheduler"), );
```
Really? TaskScheduler is never nullptr in plugin contexts

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+    connect(pcrTask, SIGNAL(si_stateChanged()), SLOT(sl_onFindTaskFinished()));
+
+    ts->registerTopLevelTask(pcrTask);
+    pbStart->setEnabled(false);
```
Does this mean that you cannot start 2 searches in parallel within the same ADV?

DmitriiSukhomlinov replied:
Yes, it's not good because we have the only result table

yalgaer replied:
Ok, I will test that it works correctly and will resolve

yalgaer replied:
I checked the UI: If I 
1) Start the task
2) Close the tab while task is active
3) Open the tab again

I see that the button is enabled again.

You can fix it right in this PR or create a separate bug for it.  This bug is not a blocker for "merge-to-master", but is needed to be fixed to include the plugin into "release".

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+
+void PCRPrimerDesignForDNAAssemblyOPWidget::sl_updateParametersRanges() {
+    auto seqLength = annDnaView->getActiveSequenceContext()->getSequenceLength();
+    const auto& parametersMinMaxSpinBoxesKeys = parametersMinMaxSpinBoxes.keys();
```
Not a reference, but a shared copy.

DmitriiSukhomlinov replied:
ОК

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+
+        minSb->setMaximum(maxSb->value() - 1);
+        maxSb->setMinimum(minSb->value() + 1);
+        maxSb->setMaximum(seqLength);
```
set maxSb->setMaximum(seqLength); first. 
Otherwise in your code minSb may have a value above maxSb if sequence length was reduced

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+    SAFE_POINT_OP(os, );
+    AnnotationTableObject *resultsTableObject = new AnnotationTableObject(PCR_TABLE_OBJECT_NAME, localDbiRef);
+    QSet<QString> excludeList;
+    for (Document* d : AppContext::getProject()->getDocuments()) {
```
qAsConst

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,519 @@
+    ADVSequenceObjectContext *sequenceContext = annDnaView->getActiveSequenceContext();
+    CHECK(sequenceContext != nullptr, )
+    const DNAAlphabet *alphabet = sequenceContext->getAlphabet();
+    SAFE_POINT(alphabet != nullptr, L10N::nullPointerError("Alphabet"), )
```
alphabet can't be null here

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyPlugin.cpp:
```diff
@@ -0,0 +1,67 @@
+
+    foreach (XMLTestFactory *f, l->qlist) {
+        bool res = xmlTestFormat->registerTestFactory(f);
+        assert(res);
```
Do not use assert(). We have SAFE_POINT. Or just ignore the result  - if test is not registered it will fail.

DmitriiSukhomlinov replied:
OK

yalgaer replied:
The change breaks the build.  Add a return value to the SAFE_POINT.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/UnwantedStructuresInBackboneDialog.cpp:
```diff
@@ -0,0 +1,53 @@
+    QString sequenceDeclension;
+    switch (sequencesCandidatesNumber) {
+        case 0:
+            sequenceDeclension = QObject::tr("are no sequences");
```
Make the dialog Q_OBJECT and do not call tr() on QObject

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/options_panel/GeneratedSequencesTable.cpp:
```diff
@@ -0,0 +1,75 @@
+void GeneratedSequencesTable::updateSequenceList(const QString& filter) {
+    clearContents();
+    QList<QString> filtred;
+    for (const auto& sequence : sequences) {
```
qAsConst

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/options_panel/ResultTable.cpp:
```diff
@@ -0,0 +1,113 @@
+    setHorizontalHeaderLabels(QStringList() << tr("Fragment") << tr("Region"));
+    setSelectionBehavior(QAbstractItemView::SelectRows);
+    setSelectionMode(QAbstractItemView::SingleSelection);
+    connect(selectionModel(), SIGNAL(selectionChanged(const QItemSelection &, const QItemSelection &)), SLOT(sl_selectionChanged()));
```
nit: use function pointers instead of SIGNAL/SLOT macros

DmitriiSukhomlinov replied:
I don't get why this is preferable

yalgaer replied:
SIGNAL and SLOTS are checked at runtime. Function pointers are checked at compile time.

DmitriiSukhomlinov replied:
This is the first time I here this thing and really not sure that it's true. `SIGNAL` and `SLOTS` are pre-processor macro definitions and they being replaced with the C++ code even before the compilation process starts (also, I suppose, they are being replaced with the same function pointers you want me to replace now). 

This is very strange thing you are talking about and at least I wanna read some proofs.

yalgaer replied:
OK, please read the proofs you need and fix the code. You may start with QT docs.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/options_panel/ResultTable.cpp:
```diff
@@ -0,0 +1,113 @@
+    int index = 0;
+    int row = 0;
+    setRowCount(MAXIMUM_ROW_COUNT);
+    for (const U2Region &region : _currentProducts) {
```
qAsConst

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/options_panel/ResultTable.cpp:
```diff
@@ -0,0 +1,113 @@
+    int row = 0;
+    setRowCount(MAXIMUM_ROW_COUNT);
+    for (const U2Region &region : _currentProducts) {
+        if (region != U2Region()) {
```
Use isEmpty()?

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/options_panel/ResultTable.cpp:
```diff
@@ -0,0 +1,113 @@
+void ResultTable::sl_selectionChanged() {
+    Annotation *selectedAnnotation = getSelectedAnnotation();
+    if (selectedAnnotation != nullptr) {
+        for (ADVSequenceObjectContext *context : associatedView->getSequenceContexts()) {
```
qAsConst

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp:
```diff
@@ -0,0 +1,193 @@
+}
+
+Document *ExtractPrimerTask::takeResult() {
+    CHECK(nullptr != result, nullptr);
```
nit: do not use Yoda style

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp:
```diff
@@ -0,0 +1,193 @@
+    int resultIndex = PCRPrimerDesignForDNAAssemblyTask::FRAGMENT_INDEX_TO_NAME.indexOf(settings.fragmentName);
+    bool isForvard = resultIndex % 2 == 0;
+
+    U2Region fragmentRegion = U2Region(0, productSequence.length());
```
nit: copy operator is not needed.

DmitriiSukhomlinov replied:
The compiler isn't a dumb, it could handle, that there are an extra operation here and fix it on it's own. It won't call a copy operator in this case, the usual constructor only. But OK, I'll replace it

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp:
```diff
@@ -0,0 +1,193 @@
+
+    QList<SharedAnnotationData> annotations;
+    int resultIndex = PCRPrimerDesignForDNAAssemblyTask::FRAGMENT_INDEX_TO_NAME.indexOf(settings.fragmentName);
+    bool isForvard = resultIndex % 2 == 0;
```
nit: Forvard -> Forward

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp:
```diff
@@ -0,0 +1,193 @@
+    DNASequence primerSequence("", "");
+    DbiConnection connection(settings.sequenceRef.dbiRef, stateInfo);
+    CHECK_OP(stateInfo, primerSequence);
+    SAFE_POINT_EXT(NULL != connection.dbi, setError(L10N::nullPointerError("DBI")), primerSequence);
```
nit: yoda style + NULL

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp:
```diff
@@ -0,0 +1,193 @@
+
+U2::Task::ReportResult ExtractPrimerAndOpenDocumentTask::report() {
+    if (extractTask->isCanceled()) {
+        QFile::remove(settings.outputFileUrl);
```
you already removed it in onSubTaskFinished ?

DmitriiSukhomlinov replied:
@rasputinkirill check this stuff, please, where it should be?

rasputinkirill replied:
This is "if task cancelled" branch.

yalgaer replied:
Ok, this is a valid point. However I would prefer having a single place the file is deleted. Not a blocker so. Resolving.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp:
```diff
@@ -0,0 +1,193 @@
+void ExtractPrimerAndOpenDocumentTask::prepareUrl() {
+    // generate file name
+    const QString fragmentNameInLowerCase = settings.fragmentName.toLower().replace(" ", "_");
+    QString fileName = settings.originalSequenceFileName + "_" + fragmentNameInLowerCase;
```
What if fragment name contains other illegal symbols?

DmitriiSukhomlinov replied:
What symbols, for example?

yalgaer replied:
Any URL usafe symbols that must be escaped. Example: ":" or "/"

DmitriiSukhomlinov replied:
@rasputinkirill check this one

rasputinkirill replied:
`fragmentName ` came from product annotation name which is safe, we do not need such kind of checks.

yalgaer replied:
Do you mean that I can't create annotations with a name that will be an illegal file name? I can create an annotation like "a:b/c?d"

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp:
```diff
@@ -0,0 +1,193 @@
+
+    // reserve file
+    QFile file(settings.outputFileUrl);
+    file.open(QIODevice::WriteOnly);
```
Check result. May be read-only

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.h:
```diff
@@ -0,0 +1,95 @@
+    ExtractPrimerTask(const ExtractPrimerTaskSettings &settings);
+    ~ExtractPrimerTask();
+    /*
+    void prepare();
```
nit: remove unused code

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.h:
```diff
@@ -0,0 +1,95 @@
+    void prepare();
+    QList<Task *> onSubTaskFinished(Task *subTask);
+    */
+    void run();
```
add override

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.h:
```diff
@@ -0,0 +1,95 @@
+
+    ExtractPrimerTaskSettings settings;
+    Document *result = nullptr;
+    LoadDocumentTask *loadTask;
```
init all pointers with nullptr?

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.h:
```diff
@@ -0,0 +1,95 @@
+    ExtractPrimerAndOpenDocumentTask(const ExtractPrimerTaskSettings &settings);
+
+    // Task
+    void prepare();
```
Add override

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/FindUnwantedIslandsTask.cpp:
```diff
@@ -0,0 +1,136 @@
+    /**
+     * Index of the left nucleotide in the searching area.
+     */
+    int leftNucleotide = searchArea.startPos;
```
Cast explicitly from qint64 to int and add a comment why this cast is safe

DmitriiSukhomlinov replied:
Why is this cast safe? Cuz we limit the sequence with 2GB?

yalgaer replied:
any cast from int64 to int32 should be explicit and commented otherwise a reader will not know if this is a bug or a feature.
This case is valid here if you limit the sequence size to 2Gb from the very beginning.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/FindUnwantedIslandsTask.cpp:
```diff
@@ -0,0 +1,136 @@
+        bool isIsland = hasUnwantedConnections(islandCandidate);
+        if (isIsland) {
+            //The obvious limit - we don't need regions which couldn't fit the primer
+            if (lengthBetweenIslands != 0/*>= overlap.minValue*/) {
```
Remove such comments or fix the code if the comment is valid

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/FindUnwantedIslandsTask.cpp:
```diff
@@ -0,0 +1,136 @@
+            lengthBetweenIslands++;
+        }
+        leftNucleotide++;
+        stateInfo.setProgress(100 - (double(rightNucleotide - leftNucleotide) / double(rightNucleotide - progressZeroPoint)) * 100);
```
According to the code above progressZeroPoint may be equal to rightNucleotide -> division by zero

DmitriiSukhomlinov replied:
Well actually, it's impossible, cuz the region length is 1 a least. But I'll fix this, just in case

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/FindUnwantedIslandsTask.cpp:
```diff
@@ -0,0 +1,136 @@
+                                                                             UNWANTED_MAX_LENGTH);
+
+    //TODO: hairpins
+    //TODO: find out if hetero-dimers are required
```
Is the algorithm complete for use in production?

DmitriiSukhomlinov replied:
Not hairpins

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp:
```diff
@@ -0,0 +1,610 @@
+};
+
+PCRPrimerDesignForDNAAssemblyTask::PCRPrimerDesignForDNAAssemblyTask(const PCRPrimerDesignForDNAAssemblyTaskSettings& _settings, const QByteArray& _sequence)
+    : Task("PCR Primer Design For DNA Assembly Task", TaskFlags_FOSCOE | TaskFlag_ReportingIsSupported | TaskFlag_ReportingIsEnabled),
```
translate task name

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp:
```diff
@@ -0,0 +1,610 @@
+namespace U2 {
+
+const QStringList PCRPrimerDesignForDNAAssemblyTask::FRAGMENT_INDEX_TO_NAME = {
+    QObject::tr("A Forward"),
```
Do not run tr() in static context. Translator may not be loaded yet.

DmitriiSukhomlinov replied:
This could be not trunslated

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp:
```diff
@@ -0,0 +1,610 @@
+
+void PCRPrimerDesignForDNAAssemblyTask::prepare() {
+    auto prepareLoadDocumentTask = [this](const QString& url) -> LoadDocumentTask* {
+        QList<FormatDetectionResult> formats = DocumentUtils::detectFormat(url);
```
This whole function can be replaced with LoadDocumentTask::getDefaultLoadDocTask ?

DmitriiSukhomlinov replied:
I suppose, yes

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp:
```diff
@@ -0,0 +1,610 @@
+    // A and B1 are pretty the same, except the fact that B1 has some depencedcies on B2 and B3
+    // So, the first time we find good B1 pair they are also the A pair. But if B1 doesn't fit to B2 or B3, we need to continue searching B1, which will be fit to B2 and B3
+    bool aWasNotFoundYet = true;
+    for (const auto& regionBetweenIslandsForward : regionsBetweenIslandsForward) {
```
qAsConst

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp:
```diff
@@ -0,0 +1,610 @@
+    // So, the first time we find good B1 pair they are also the A pair. But if B1 doesn't fit to B2 or B3, we need to continue searching B1, which will be fit to B2 and B3
+    bool aWasNotFoundYet = true;
+    for (const auto& regionBetweenIslandsForward : regionsBetweenIslandsForward) {
+        stateInfo.setProgress(((double)regionsBetweenIslandsForward.indexOf(regionBetweenIslandsForward) / (double)regionsBetweenIslandsForward.size()) * 100);
```
Really? Why not using indexed for loop instead?

DmitriiSukhomlinov replied:
Got it

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp:
```diff
@@ -0,0 +1,610 @@
+    bool aWasNotFoundYet = true;
+    for (const auto& regionBetweenIslandsForward : regionsBetweenIslandsForward) {
+        stateInfo.setProgress(((double)regionsBetweenIslandsForward.indexOf(regionBetweenIslandsForward) / (double)regionsBetweenIslandsForward.size()) * 100);
+        int progressStage = 100 / regionsBetweenIslandsForward.size();
```
Double? Move out of the loop?

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp:
```diff
@@ -0,0 +1,610 @@
+                updatePrimerRegion(b1ForwardCandidatePrimerEnd, b1ForwardPrimerLength);
+                // ... and check the primer again
+                continue;
+            } else {
```
nit: else  is not needed

DmitriiSukhomlinov replied:
The offset is more convenient with else

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp:
```diff
@@ -0,0 +1,610 @@
+
+void PCRPrimerDesignForDNAAssemblyTask::findB1ReversePrimer(const QByteArray& b1ForwardCandidatePrimerSequence) {
+    // Very the same algorithm, but for reverse regions etween islands
+    for (const auto& regionBetweenIslandsReverse : regionsBetweenIslandsReverse) {
```
qAsConst

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.h:
```diff
@@ -0,0 +1,112 @@
+        B2,
+        B3
+    };
+    void findSecondaryForwardReversePrimers(SecondaryPrimer type);
```
Document side effects

DmitriiSukhomlinov replied:
What side effects?

yalgaer replied:
The reason by the method is not const. Ok, this is not a big issue, resolving

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp:
```diff
@@ -0,0 +1,610 @@
+        }
+    }
+    generateUserPrimersReports();
+    stateInfo.setProgress(100);
```
Not needed: set by the scheduler automatically

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTaskTest.cpp:
```diff
@@ -0,0 +1,127 @@
+        stateInfo.setError(GTest::tr("context not found %1").arg(docName));
+        return;
+    }
+    U2SequenceObject *dnaso = (U2SequenceObject *)loadedDocument->findGObjectByName(seqName);
```
use qobject_cast

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTaskTest.cpp:
```diff
@@ -0,0 +1,127 @@
+
+void GTest_PCRPrimerDesignForDNAAssemblyTaskTest::prepare() {
+    Document *loadedDocument = getContext<Document>(this, docName);
+    if (loadedDocument == nullptr) {
```
nit: use CHECK_EXT

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTaskTest.cpp:
```diff
@@ -0,0 +1,127 @@
+        stateInfo.setError(GTest::tr("Sequence %1 not found").arg(seqName));
+        return;
+    }
+    task = new PCRPrimerDesignForDNAAssemblyTask(settings, dnaso->getWholeSequenceData(stateInfo));
```
Check stateInfo before task is created

DmitriiSukhomlinov replied:
OK

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/utils/PCRPrimerDesignTaskReportUtils.cpp:
```diff
@@ -0,0 +1,283 @@
+
+// Returns error message in html.
+static QString errMsg() {
+    return QObject::tr("<p>Error, see log.</p>");
```
Do not use global namespace (QObject) for plugin local messages. Use plugin class name if there is no better choice.

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer replied:
This place is fixed. But there are other like this. Try
`grep -r "QObject::tr" .`

evelinabis replied:
@yalgaer, please see changes.

yalgaer replied:
Fixed. Thank you.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/utils/PCRPrimerDesignTaskReportUtils.cpp:
```diff
@@ -0,0 +1,283 @@
+////////////////////////////////////////////////////UserPrimersTable////////////////////////////////////////////////////
+QString PCRPrimerDesignTaskReportUtils::UserPrimersTable::getRow(const QList<bool> &row) const {
+    QString ans;
+    for (bool cell : row) {
```
qAsConst

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/utils/PCRPrimerDesignTaskReportUtils.cpp:
```diff
@@ -0,0 +1,283 @@
+
+QString PCRPrimerDesignTaskReportUtils::UserPrimersTable::getTable() const {
+    QString header;
+    for (const QString &h : headers) {
```
qAsConst

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/utils/PCRPrimerDesignTaskReportUtils.h:
```diff
@@ -0,0 +1,106 @@
+};
+
+class PCRPrimerDesignTaskReportUtils::UserPrimersTable {
+    const QString name;  // Name of table.
```
nit: use members/private down in the class declaration

evelinabis replied:
@DmitriiSukhomlinov, @yalgaer, please see changes.

yalgaer reviewed (changes requested):
These are the most comments for different minor code-issues I found.

Next I will check the PR in runtime (need 1-2 days for it)

Commit:
UGENE-7243. PCR Primer Design for DNA assembly.

Changes by review.

yalgaer reviewed (changes requested):
Resolved all fixed issues. Will review again after the next commit

DmitriiSukhomlinov committed:
Review fix

yalgaer reviewed (changes requested):
Reviewed all changes. Some open issues are  left. Please re-request the review again once they are processed.

rasputinkirill committed:
Update test_0001.xml

evelinabis committed:
UGENE-7243. PCR Primer Design for DNA assembly.

Changes by review:
pointer is replaced with alphabet id in FilterVisitor,
QObject::tr has been removed.

evelinabis committed:
UGENE-7243. PCR Primer Design for DNA assembly.

Fixed a crash associated with calling tr() for a static variable.

evelinabis committed:
UGENE-7243. PCR Primer Design for DNA assembly.

Fixed Linux build related to SAFE_POINT with two arguments.

evelinabis commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTaskTest.cpp:
```diff
@@ -0,0 +1,126 @@
+
+void GTest_PCRPrimerDesignForDNAAssemblyTaskTest::prepare() {
+    Document *loadedDocument = getContext<Document>(this, docName);
+    CHECK_EXT(loadedDocument != nullptr, stateInfo.setError(GTest::tr("context not found %1").arg(docName)), );
```
Is translation required here? The same 3 lines below.

yalgaer replied:
No, there is not need in translation of tests code.

evelinabis commented on src/plugins/pcr_primer_design_for_dna_assembly/transl/russian.ts:
```diff
@@ -0,0 +1,703 @@
+    </message>
+    <message>
+        <location filename="../src/PCRPrimerDesignForDNAAssemblyOPWidget.ui" line="269"/>
+        <source>Reverse 3&apos;</source>
+        <translation>Обратный 5&apos;</translation>
+    </message>
+    <message>
+        <location filename="../src/PCRPrimerDesignForDNAAssemblyOPWidget.ui" line="276"/>
+        <source>Reverse 5&apos;</source>
+        <translation>Обратный 3&apos;</translation>
```
@DmitriiSukhomlinov, this is mistake?

evelinabis commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/FindUnwantedIslandsTask.cpp:
```diff
@@ -0,0 +1,134 @@
+    taskLog.details(tr("Searching of unwanted islands and areas between them "
+                       "in the region \"%1\" (+ %2 nucleotides to the %3, deep into the amplified fragment) has been started")
+                       .arg(regionToString(searchArea)).arg(possibleOverlap).arg(isComplement ? "left" : "right"));
+    taskLog.details(tr("The following unwanted parametes are used. Free Gibbs energy: %1 kj/mol, "
```
kj or kcal?

evelinabis committed:
Translations for PCR Primer Design for DNA assembly (#654)

* UGENE-7243. PCR Primer Design for DNA assembly.

Translations corrected.

DmitriiSukhomlinov committed:
Widget width fix

yalgaer commented:
Closing as inactive. Please reopen and ping me when the patch is ready for review.

evelinabis commented on src/plugins/pcr_primer_design_for_dna_assembly/src/options_panel/UserPimerLineEdit.cpp:
```diff
@@ -0,0 +1,216 @@
+
+QColor UserPimerLineEdit::getColorForInsertedText(int pos) const {
+    QColor resultColor;
+    int primerSize = qMax(0, lastTextSize - (leftEndSize + rightEndSize));
+    if (changeLeftEndColor || (0 <= pos && pos < leftEndSize)) {
+        resultColor = Qt::red;
+    } else if (!changeLeftEndColor && !changeRightEndColor && leftEndSize <= pos && pos <= leftEndSize + primerSize) {
```
When you reopen the tab, the backlight breaks. I suggest using `text().size()` instead of `lastTextSize` , and replace `pos <= leftEndSize + primerSize` with `pos < leftEndSize + primerSize` before `Qt::black`.



I also don't know if the current behavior is correct: if you select multiple letters, whether they are the same color or not, and insert some other letters instead, they will be inserted in black. For example:
![image](https://user-images.githubusercontent.com/39720063/150507285-4fe33bfc-6cac-42c3-8ad4-5d3ce6b83148.png)

evelinabis commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.ui:
```diff
@@ -0,0 +1,1047 @@
+           </property>
+           <property name="columnCount">
+            <number>1</number>
+           </property>
```
On my screen, the word "Последовательность" does not fit in the column completely (![image](https://user-images.githubusercontent.com/39720063/150520162-dd6d5c5c-2e4d-4e9a-bea7-6ccdeba6db11.png)). I suggest increasing the column width to at least 139.
```
<attribute name="horizontalHeaderDefaultSectionSize">
  <number>139</number>
</attribute>
```

evelinabis commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,512 @@
+    // Right area: from "50% of sequence length" to "60% of sequence length"
+    sbLeftAreaStart->setValue(1);
+    sbLeftAreaEnd->setValue(seqLength / 10);
+    sbRightAreaStart->setValue(seqLength / 10 * 5);
+    sbRightAreaEnd->setValue(seqLength / 10 * 6);
```
Loss of accuracy, or did the customer want to have starting coordinates that are multiples of five? Use `static_cast<double>(seqLength)` or `seqLength * 0.5`.

evelinabis commented on src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/FindUnwantedIslandsTask.cpp:
```diff
@@ -0,0 +1,136 @@
+     * Index of the right nucleotide in the searching area.
+     * @settings.overlapLength.maxValue is the area extending deep into the amplified fragment.
+     */
+    const int rightNucleotide = searchArea.endPos() + possibleOverlap;
```
EndPos is used here, but here https://github.com/ugeneunipro/ugene/blob/771b46aac08a5d4832a746b87cac3dd4cbe00383/src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp#L108 endPos-1 is used.

evelinabis commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -0,0 +1,512 @@
+    settings.bachbone3Length = sbBackbone3Length->value();
+
+    settings.leftArea.startPos = sbLeftAreaStart->value() - 1;
+    settings.leftArea.length = sbLeftAreaEnd->value() - sbLeftAreaStart->value();
+    settings.rightArea.startPos = sbRightAreaStart->value() - 1;
+    settings.rightArea.length = sbRightAreaEnd->value() - sbRightAreaStart->value();
```
For length, replace `sbLeftAreaStart->value()` with `settings.leftArea.startPos`. The same for `rightArea`. If you did, replace `area.length >= 0` with `area.length > 0` at https://github.com/ugeneunipro/ugene/blob/59e80071b0587d3a2022382a45b1d03f05c406b9/src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp#L240
